NEP-330 extension: Build details extension (#533)
near/NEPs: neps/nep-0330.md

- `build_info`: a build details object (see type definition below) that contains all the necessary information about how the contract was built, making it possible for others to reproduce the same WASM of this contract.

```ts
type ContractSourceMetadata = {
    version: string|null, // optional, commit hash being used for the currently deployed wasm. If the contract is not open-sourced, this could also be a numbering system for internal organization / tracking such as "1.0.0" and "2.1.0".
    link: string|null, // optional, link to open source code such as a Github repository or a CID to somewhere on IPFS. e.g. https://"github.com/near-examples/nft-tutorial/tree/39f2d2646f2f60e18ab53337501370dc02a5661c"
    standards: Standard[]|null, // optional, standards and extensions implemented in the currently deployed wasm e.g. [{standard: "nep330", version: "1.1.0"},{standard: "nep141", version: "1.0.0"}].
    build_info: BuildInfo|null, // optional, details that are required for contract WASM reproducibility.
}

type Standard {
    standard: string, // standard name e.g. "nep141"
    version: string, // semantic version number of the Standard e.g. "1.0.0"
}

type BuildInfo {

>     build_environment: string, // reference to a reproducible build environment docker image, e.g. "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471" or something else pointing to the build environment. When using a Docker image as a reference, it's important to specify the digest of the image to ensure reproducibility, since a tag could be reassigned to a different image.

Reviewer: I suggest we recommend to use `docker.io/` prefix for Docker Hub images, so when anyone introduces torrent link or something, it is easier to tell them apart.

```suggestion
    build_environment: string, // reference to a reproducible build environment docker image, e.g. "docker.io/sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471" or something else pointing to the build environment. When using a Docker image as a reference, it's important to specify the digest of the image to ensure reproducibility, since a tag could be reassigned to a different image.
```

    source_code_snapshot: string, // reference to the source code snapshot that was used to build the contract, e.g. "git+https://github.com/near-DevHub/neardevhub-contract.git#335e89edec95d56a4744e7160c3fd590e41ec38e" or "ipfs://<ipfs-hash>"

> 	  contract_path: string|null, // relative path to contract folder within the source code e.g. "src/contract". Often, it is the root of the repository, so can be omitted. 
> 	  build_command: string[], // the exact command that was used to build the contract, with all the flags e.g. ["cargo", "near", "build", "no-abi"].

Reviewer 2: Fix the indentation.
```suggestion
    contract_path: string|null, // relative path to contract folder within the source code e.g. "src/contract". Often, it is the root of the repository, so can be omitted. 
    build_command: string[], // the exact command that was used to build the contract, with all the flags e.g. ["cargo", "near", "build", "no-abi"].
```

>     contract_path: string|null, // relative path to contract crate within the source code e.g. "src/contract". Often, it is the root of the repository, so can be omitted. 

Reviewer: `src/` might be misleading and readers may think they should somehow reference a file within `src/` folder (standard in Rust projects), so I suggest to use more neutral example:

```suggestion
    contract_path: string|null, // relative path to contract crate within the source code e.g. "contracts/contract-one". Often, it is the root of the repository, so can be omitted. 
```

Successful implementations of this standard will introduce a new  (`ContractSourceMetadata`) struct that will hold all the necessary information to be queried for. This struct will be kept on the contract level.

The metadata will include optional fields:

- `version`: a string that references the specific commit hash or version of the code currently deployed on-chain. If the link is to a GitHub repo URL, this is crucial, as it is needed to make the contract reproducible.

> - `link`: a string that references the link to the open-source code. This can be anything such as Github or a CID to somewhere on IPFS.

Reviewer: @lachlanglen Would this approach of a "frozen" link work for you? I believe it should be a good replacement for a `commit_hash` field which assumes that the code is stored on a public git provider.

> - `link`: a string that references the link to the open-source code. This can be anything such as Github or a CID to somewhere on IPFS.

Author: > Thus, I would move this build-related source code link to the `build_details` and don't touch the top-level `link` or `version` at all.

I agree. I appreciate the approach of using a "frozen" link, as it appears to resolve all the issues related to the "link" and "version" fields. The reason I didn't add it initially was to avoid confusion for anyone reviewing the metadata. Therefore, the name of the new field should be distinctly different from "link". In my opinion, "source_snapshot" fits well.

> - `link`: a string that references the link to the open-source code. This can be anything such as Github or a CID to somewhere on IPFS.

Author: ```ts
type BuildInfo {
    build_environment_ref: string, // reference to a reproducible build environment docker image, e.g. "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471" or something else pointing to the build environment.
    source_code_snapshot: string, // reference to the source code snapshot that was used to build the contract, e.g. "git+https://github.com/near-DevHub/neardevhub-contract.git#335e89edec95d56a4744e7160c3fd590e41ec38e" or "ipfs://<ipfs-hash>"
    contract_path: string|null, // relative path to contract folder within the source code e.g. "src/contract". Often, it is the root of the repository, so can be omitted. 
    build_command: string[], // the exact command that was used to build the contract, with all the flags e.g. ["cargo", "near", "build", "no-abi"].
}
```

Now using `source_code_snapshot` in 18968ea09c965845b92ab7fdb39a8a1df319b7cd

- `standards`: a list of objects (see type definition below) that enumerates the NEPs supported by the contract. If this extension is supported, it is advised to also include NEP-330 version 1.1.0 in the list (`{standard: "nep330", version: "1.1.0"}`).
- `build_details`: a build details object (see type definition below) that contains all the necessary information about how the contract was built, making it possible for others to reproduce the same WASM of this contract.

```ts
type ContractSourceMetadata = {
  version: string|null, // optional, commit hash being used for the currently deployed wasm. If the contract is not open-sourced, this could also be a numbering system for internal organization / tracking such as "1.0.0" and "2.1.0".
  link: string|null, // optional, link to open source code such as a Github repository or a CID to somewhere on IPFS.
  standards: Standard[]|null, // optional, standards and extensions implemented in the currently deployed wasm e.g. [{standard: "nep330", version: "1.1.0"},{standard: "nep141", version: "1.0.0"}].

>   build_details: BuildDetails|null, // optional, details that are required for contract WASM reproducibility.

Reviewer 3: ```suggestion
  build_info: BuildInfo|null, // optional, details that are required for contract WASM reproducibility.
```

Reviewer 3: I think _details_ sounds more like the output details from the build, not _how to_ build.

Reviewer: Thanks for providing your reasoning. I can see your point, I am fine with "info".

}

type Standard {
    standard: string, // standard name e.g. "nep141"
    version: string, // semantic version number of the Standard e.g. "1.0.0"
}

> type BuildDetails {

Reviewer 3: ```suggestion
type BuildInfo {
```

	env_image: EnvImage, // reference to a reproducible build environment image, e.g. "{tag: "sourcescan/cargo-near:0.6.0", digest: "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"}"

> 	contract_path: string|null // relative path to contract folder within the source code e.g. "src/contract". Often, it is the root of the repository, so can be omitted. 

Reviewer 3: ```suggestion
	contract_path: string|null // relative path to contract crate within the source code e.g. "contracts/contract1". Often, it is the root of the repository, so can be omitted. 
```

Author: Changed here 66db5e8f2de6c32eaa15d05c01d811b2f4ee0456.
About the path example, I am referencing to the template here: https://github.com/near/cargo-near-new-project-template.

> 	build_command: string[] // the exact command that was used to build the contract, with all the flags e.g. ["cargo", "near", "build", "no-abi"].

Reviewer 3: instead of list of strings, should we just use a string, with proper flags?

Reviewer: @robert-zaremba The reasoning behind the list of strings is that you can just pass those directly to `exec`. Having a string there would require shell parsing and may introduce unnecessary side-effects, e.g. if you would need to escape whitespace in the command, etc.

Reviewer 3: ok, fair enough.

}

> type EnvImage {

Reviewer 3: ```suggestion
type Container {
```

Reviewer 2: @robert-zaremba A container is a running instance of the image, so I believe it's still an image.

Reviewer: Originally, it was named `build_environment_ref`, and it was expected to be either `docker.io/frol/near-builder@hash-of-the-image` or something less conventional (torrent link, IPFS cid, etc)

Author: ```ts
type BuildInfo {
    build_environment_ref: string, // reference to a reproducible build environment image, e.g. "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471" or something else pointing to the build environment.
```

Now using `build_environment_ref` in 18968ea09c965845b92ab7fdb39a8a1df319b7cd

Reviewer 3: honestly, `Env` in `EnvImage` doesn't sound right. `build_environment_ref` is even more cryptic. Why not `BuilderImage` . We can also say `DockerImage` if we don't want to use _Container_

Author: I like the name 'builder_image', but I think the original idea was to make it more abstract. This way, the reference could be something other than just a Docker image, for example, a link to a CD pipeline that creates the release.

	tag: string // tag of the image e.g. "sourcescan/cargo-near:0.6.0"

> 	digest: string // a sha256 hash unique, immutable identifier for a container image, used to verify that the image under the same tag has not changed e.g. "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"

Reviewer 2: What's the function of the `digest` field? Is it just a long unique number, or is it supposed to be a hash of something specific?

Reviewer 2: @frol So, the `digest` is a hash of a _Docker_ container image on the hub.docker.com, correct?

The word "docker" should be mentioned at least once in the NEP, because it's a bit confusing.

Author: > The word "docker" should be mentioned at least once in the NEP, because it's a bit confusing.

Good point. I will clarify this to prevent confusion. Additionally, it's worth noting that the digest might differ across various registries. Therefore, the structure should also be modified as follows:
```ts
type DockerImage {
  registry: string // Docker image registry e.g. "ghcr.io", "hub.docker.com"
  tag: string // tag of the Docker image e.g. "sourcescan/cargo-near:0.6.0"
  digest: string // a sha256 hash unique, immutable identifier for a Docker image, used to verify that the image under the same tag has not changed e.g. "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"
}
```

Author: 35bbcded50a0419a5a24090bd2b0e27b8bcafed5

> 	digest: string // a sha256 hash unique, immutable identifier for a container image, used to verify that the image under the same tag has not changed e.g. "sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471"

Author: Fixed in f47d5ede90e0e542b99ac9f81b257789f21d7b4c

## Reference Implementation

> As an example, consider an NFT contract located at the relative path `nft-contract`, which was deployed using the `cargo near deploy --no-abi` environment docker image `sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471`. Its latest commit hash is `39f2d2646f2f60e18ab53337501370dc02a5661c`, and its open-source code can be found at `https://github.com/near-examples/nft-tutorial`. This contract would then include a struct with the following fields:

Reviewer 2: @Canvinus @frol @robert-zaremba  Could we suggest having a `Cargo.lock` in the contract folder?

I think this way: if the contract was hacked because of a security issue in a 3rd-party dependency, then such an issue may not be reproducible after the dependency upgrade (in the least "patch" version triplet).

Also, I checked the `nft-tutorial/nft-contract`, it has `[patch.crates-io]` which leads to a custom crate, and it may be hard to reproduce the contract with such a dependency.

Reviewer 3: Yes, good idea!

Author: Added note about Cargo.lock in 222d2602059765f1a127223840a72f2de64bb219
And changed the example repo to https://github.com/near/cargo-near-new-project-template in 35bed0fbc75783132b7ca95a11c9c51974844824

Reviewer 2: @Canvinus Thanks but Сargo.lock is still necessary because dependencies of dependencies may change.

Author: @fadeevab, I have info about this in the comment of source_code_snapshot, where do you think it is also necessary to include information about Cargo.lock

Reviewer 2: @Canvinus
Thank you for your clarification questions!

1. I see that https://github.com/near/core-contracts have `Cargo.lock`s, it could be used as an example.
2. I realize we already use  https://github.com/near/cargo-near-new-project-template: is it possible to commit `Cargo.lock` there?
3. I would suggest moving that sentence from the code comment to the specification (somewhere before the reference implementation).

```
### Cargo.lock

It is important to have `Cargo.lock` inside the source code snapshot to ensure reproducibility. Example: https://github.com/near/core-contracts.
```
👇
![image](https://github.com/near/NEPs/assets/5967447/815f3b2e-b511-47d7-b35b-a23ea664befc)

Author: Great that most of this project is still compiling, thanks to Cargo.lock: https://github.com/near/cargo-near-new-project-template. But I thought, isn't that a problem that they are using an outdated SDK that is not even capable of having contract source metadata? Wouldn't this be confusing to use as a reference?

Author: Added new section `Ensuring WASM Reproducibility` c06b054f4ec19e77ce92df2fe30a4bab5b8e0ea7 and moved info out of comments into it

Reviewer 2: > Added new section `Ensuring WASM Reproducibility` [c06b054](https://github.com/near/NEPs/commit/c06b054f4ec19e77ce92df2fe30a4bab5b8e0ea7) and moved info out of comments into it

Awesome!

>  Wouldn't this be confusing to use as a reference?

An alternative is to add `Cargo.lock` to https://github.com/near/cargo-near-new-project-template and use it as a reference. It would be the best option probably...

```ts
type ContractSourceMetadata = {
    version: "1.0.0",
    link: "https://github.com/near-examples/nft-tutorial/tree/39f2d2646f2f60e18ab53337501370dc02a5661c",
    standards: [
        {
            standard: "nep330",
            version: "1.1.0"
        },
        {
            standard: "nep171",
            version: "1.0.0"
        },
        {
            standard: "nep177",
            version: "2.0.0"
        }
    ],
    build_info: {

>         build_environment: "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471",

Reviewer 2: @Canvinus  Also, you should add `docker.io` here:
```suggestion
        build_environment: "docker.io/sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471",
```

## Reference Implementation

As an example, consider an NFT contract located at the relative path `nft-contract`, which was deployed using the `cargo near deploy --no-abi` environment image `sourcescan/cargo-near:0.6.0`. Its latest commit hash is `39f2d2646f2f60e18ab53337501370dc02a5661c`, and its open-source code can be found at `https://github.com/near-examples/nft-tutorial`. This contract would then include a struct with the following fields:

```ts
type ContractSourceMetadata = {
  version: "39f2d2646f2f60e18ab53337501370dc02a5661c",

>   link: "https://github.com/near-examples/nft-tutorial",

Author: Updated here 04c54a5040d4f73e76b15087022c652d82823693

            standard: "nep177", 
            version: "2.0.0"
        }
    ],
    build_info: {

>         build_environment: "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471",

Reviewer 2: @Canvinus  And here:
```suggestion
        build_environment: "docker.io/sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471",
```

Author: Yes, thanks

>     build_info: {
> 	      build_environment_ref: "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471",
>         source_code_snapshot: "git+https://github.com/near-examples/nft-tutorial.git#39f2d2646f2f60e18ab53337501370dc02a5661c",
> 	      contract_path: "nft-contract",
> 	      build_command: ["cargo", "near", "deploy", "--no-abi"]
> 	  }

Reviewer 2: Fix the indentation 
```suggestion
    build_info: {
        build_environment_ref: "sourcescan/cargo-near@sha256:bf488476d9c4e49e36862bbdef2c595f88d34a295fd551cc65dc291553849471",
        source_code_snapshot: "git+https://github.com/near-examples/nft-tutorial.git#39f2d2646f2f60e18ab53337501370dc02a5661c",
        contract_path: "nft-contract",
        build_command: ["cargo", "near", "deploy", "--no-abi"]
    }
```

Reviewer 2: I like flatting the structure! 👌
The `_ref` suffix is a bit out of the overall naming convention, could it be like a full word `_reference`, or maybe drop it, e.g. `build_environment`? Or `build_environment_link` (cause we already use the `link` word in a different place)?

Reviewer 3: let's drop it.

Author: dropped `_ref` in aa84739feadf6b04bbfd7fdd06102b6d189367c8
